src/tools/tests/schema2kotlin-test.ts

```diff
@@ -219,10 +250,122 @@ describe('schema2kotlin', () => {
+        'typealias T_H3 = AbstractT.TInternal1',
+      ]
+    ));
     async function assertSchemaAliases(manifest: string, expectedAliases: string[]) {
       await assertComponent(manifest, ({typeAliases}) => typeAliases.sort(), expectedAliases);
     }
   });
+  describe('Kotlin Generator', () => {
+    it('generates entity with public constructor', async () => await assertClassDefinition(
+      `particle T
+         h1: reads Thing {num: Number}`,
+      `@Suppress("UNCHECKED_CAST")
+    class T_H1(
+        num: Double = 0.0,
+        entityId: String? = null,
+        creationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP,
+        expirationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP
+    ) : EntityBase("T_H1", SCHEMA, entityId, creationTimestamp, expirationTimestamp)`
+    ));
+    it('generates variable entity with private constructor', async () => await assertClassDefinition(
+      `particle T
+         h1: reads ~a with {num: Number}
+         `,
+      `@Suppress("UNCHECKED_CAST")
+    class T_H1 private constructor(
+        num: Double = 0.0,
+        entityId: String? = null,
+        creationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP,
+        expirationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP
+    ) : EntityBase("T_H1", SCHEMA, entityId, creationTimestamp, expirationTimestamp)`
+    ));
+    it('generates copy and mutate by entity fields', async () => await assertCopyMethods(
+      `particle T
+         h1: reads Thing {num: Number}`,
+      `/**
+         * Use this method to create a new, distinctly identified copy of the entity.
+         * Storing the copy will result in a new copy of the data being stored.
+         */
+        fun copy(num: Double = this.num) = T_H1(num = num)
+        /**
+         * Use this method to create a new version of an existing entity.
+         * Storing the mutation will overwrite the existing entity in the set, if it exists.
+         */
+        fun mutate(num: Double = this.num) = T_H1(
+            num = num,
+            entityId = entityId,
+            creationTimestamp = creationTimestamp,
+            expirationTimestamp = expirationTimestamp
+        )`
+    ));
+    // TODO(alxr): Why do we omit the docstring for Wasm?
+    it('generates only copy method by entity fields for wasm', async () => await assertCopyMethodsForWasm(
+      `particle T
+         h1: reads Thing {num: Number}`,
+      `
+        fun copy(num: Double = this.num) = T_H1(num = num)
+        `
+    ));
+    it('copies underlying data for type variables.', async () => await assertCopyMethods(
+      `particle T
+         h1: reads ~a with Thing {num: Number}`,
+      `/**
+         * Use this method to create a new, distinctly identified copy of the entity.
+         * Storing the copy will result in a new copy of the data being stored.
+         */
+        fun copy(num: Double = this.num) = T_H1(num = num)
+            .also { this.copyInto(it) }
```

Comment:
here and below, won't the `this.copyInto(it)` override the value provided as the argument to the copy (or mutate) method?

If I hold an entity with the `num = 7`, and I do `entity.copy(42)`, wouldn't it:
1) Create an entity with num = 42.
2) Copy over all the values from the previoius entity, overriding 42 with 7?

Looks like something we could have a test for! :) It would be the first time we would be writing tests for code generated code, but I think it would actually be good to do (if a bit funky to set up) - WDYT?

Reply:
I agree, this is a great candidate to exercise the generated code. 

Yeah, as written, this definitely causes an overwrite bug. 

Maybe there's a clean Kotlin extension function that will solve this for me, I'll look around.